src/Middleware/RequestThrottling/src/RequestThrottlingMiddleware.cs

```diff
@@ -49,19 +47,17 @@ public RequestThrottlingMiddleware(RequestDelegate next, ILoggerFactory loggerFa
         /// <returns>A <see cref="Task"/> that completes when the request leaves.</returns>
         public async Task Invoke(HttpContext context)
         {
-            Interlocked.Increment(ref _queuedRequests);
+            var waitInQueueTask = _queuePolicy.TryEnterAsync();
```

Comment:
Nice little perf improvement here 🎈

Reply:
Based on @halter73 's [comment](https://github.com/aspnet/AspNetCore/pull/11293#discussion_r296336437), I changed `Task<bool>` to a resettable `ValueTask<bool>` to see the impacts, and if there's a plan to convert to `ValueTask<bool>` I think this would throw an exception, since you can't access the property of a `ValueTask` after it's resetted.

Reply:
@Kahbazi The code in the middleware here is good, what would need to switch is the code in the individual policies. I hit some of problems you describe when I tried switching to `ValueTask`; specifically resetting a `ValueTask` to immediately use it for another request is a bad idea. One approach that I tried was having n+1 `ValueTasks` and resetting not the task to be immediately used, but the task that would be needed by the next request if the queue were still full. I wasn't able to get a satisfactory & reliable implementation, but may revisit if I have time later.